### PR TITLE
CORE-9517: Add condition to reply

### PR DESCRIPTION
**Pull request by yift-r3**
Runtime pull request in https://github.com/corda/corda-runtime-os/pull/3311

This change should change the registration retry envelope. It will now contain a timestamp on when to retry again (to make two different duration depending on the cause of the retry) and an enum about the condition in which we should retry the request.

### CODE REVIEW

File: data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc [hunk 1]
```diff
@@ -17,8 +17,20 @@
       }
     },
     {
-      "name": "lastFailedOn",
-      "doc": "The time in which the request had last failed.",
+      "doc": "The condition to check before retring the request.",
```
Comment (blsemo) [inline data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc]:
```suggestion
      "doc": "The condition to check before retrying the request.",
```

File: data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc [hunk 2]
```diff
@@ -17,8 +17,20 @@
...
+        "name": "RetryCondition",
+        "symbols": [
+          "NO_CONDITION",
+          "IS_STATE_SENT_TO_MGM"
+        ]
+      }
+    },
+    {
+      "name": "retryAfter",
+      "doc": "The time after which the reqeust should retry.",
```
Comment (blsemo) [inline data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc]:
```suggestion
      "doc": "The time after which the request should be retried.",
```

File: data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc [hunk 3]
```diff
@@ -2,27 +2,38 @@
...
-        "type": "long",
-        "logicalType": "timestamp-millis"
-      }
+      "name": "cause",
+      "doc": "The cause the request might need to be retried.",
+      "type": [{
+        "type": "record",
+        "name": "RetriableFailure",
+        "fields": [{
+            "name": "numberOfRetriesToGo",
```
Comment (dimosr) [inline data/avro-schema/src/main/resources/avro/net/corda/data/membership/async/request/MembershipAsyncRequestState.avsc]:
```suggestion
            "name": "numberOfRemainingRetries",
```

File: data/config-schema/src/main/resources/net/corda/schema/configuration/membership/1.0/corda.membership.json [hunk 4]
```diff
@@ -37,6 +37,13 @@
           "description": "The maximum duration in minutes to wait for a verify member request.",
           "$ref": "#/$defs/ttl",
           "default": 20
+        },
+        "waitForMgmSession": {
+          "description": "The maximum duration in minutes to wait for a member to establish connection with the MGM.",
```
Comment (dimosr) [inline data/config-schema/src/main/resources/net/corda/schema/configuration/membership/1.0/corda.membership.json]:
```suggestion
          "description": "The maximum duration in minutes to wait for a member to establish connection with the MGM. This only has effect on time waiting for an MGM to establish a session with a member as part of processing a registration request.",
```
we should clarify it doesn't have an effect on all communication with the MGM.

**Review by dimosr — DISMISSED** [review]

**Review by dimosr — DISMISSED** [review]

**Review by dimosr — APPROVED** [review]

**Pull request merged**